arangod/RestHandler/RestReplicationHandler.cpp

```diff
@@ -3319,11 +3319,39 @@ void RestReplicationHandler::handleCommandRevisionDocuments() {
 
     for (VPackSlice entry : VPackArrayIterator(body)) {
       RevisionId rev = RevisionId::fromSlice(entry);
-      if (it.hasMore() && it.revision() < rev) {
+      if (!it.hasMore()) {
+        // If the iterator is exhausted, let's try to seek, this can happen
+        // if the client has jumped back. Normally, the iterator will never
+        // be exhausted, but we must check, since we are not even allowed
+        // to call it.revision() if the iterator is exhausted!
         it.seek(rev);
-      }
-      VPackSlice res =
-          it.hasMore() ? it.document() : velocypack::Slice::emptyObjectSlice();
+      } else if (it.revision() < rev) {
+        // This is the normal case, in most cases, the client will ask for
+        // a whole range of adjacent revisions, so we optimize the fast
+        // path for this, so we try it.next() first, since this is cheaper
+        // than a seek:
+        it.next();
+        if (!it.hasMore() || it.revision() != rev) {
```

Comment:
The API's purpose is to deliver a list of revisions for which the follower has asked. I think the API should work with whatever list of revisions it gets. The normal case is that the follower fetches missing revisisions in ascending order and we can usually assume that if it needs a revision, it will probably need the next one next, but this is not clear. The follower might have holes in its list of revisions. So we optimize performance by:
 - not moving the iterator if we happen to be in the right place, since this is a cheap check
 - using `next` for the common case that the next rev the follower asks for is the next one
 - if this is not the case, then we must do an additional seek to get the right thing.
As far as I see, this works always and has a good fast path (next vs. seek) for the most common case.